### PR TITLE
Make lookup-logic more generic

### PR DESCRIPTION
**Pull request by nielslaukens**
This Pull Request is work in progress. It has been opened to solicit feedback while it is developing.

This patch is an attempt to fix issue #664, and extending the functionality of the lookups. Currently, the `${output stack::name}` lookup is treated specially to allow it to inject dependencies. This is an attempt to make the logic more generic, so new/other lookups can also inject their dependencies.

### CODE REVIEW

File: stacker/variables.py [hunk 1]
```diff
@@ -140,21 +128,358 @@ def resolve(self, context, provider):
...
+    """
+    def value(self):
+        return NotImplementedError()
+
+    def __iter__(self):
+        return NotImplementedError()
+
+    def resolved(self):
+        """
+        :return: Whether value() will not raise an error
```
Comment (phobologic) [inline stacker/variables.py]: Can you convert these to the Google style docstrings, that's the standard we use in stacker.

**Comment by phobologic** [comment]
Sorry for the spam - I blame github :)

File: stacker/lookups/registry.py [hunk 2]
```diff
@@ -80,7 +97,7 @@ def resolve_lookups(variable, context, provider):
     return resolved_lookups
 
 
-register_lookup_handler(output.TYPE_NAME, output.handler)
+register_lookup_handler(output.TYPE_NAME, output.OutputLookup)
 register_lookup_handler(kms.TYPE_NAME, kms.handler)
```
[the remark this answers is not in the document]
Comment (nielslaukens) [reply stacker/lookups/registry.py]: yeah, but I can't get the linting to work on my machine, so I'm abusing Travis...

[the remark this answers is not in the document]
Comment (russellballestrini) [reply stacker/lookups/registry.py]: Please post error, I think I ran into this before.

[the remark this answers is not in the document]
Comment (nielslaukens) [reply stacker/lookups/registry.py]: I'm getting tons of Deprecation warnings, so the actual things I need to fix are hard to find

**Comment by nielslaukens** [comment]
This PR is, as far as I'm concerned, in a mergable state. I don't have much more time to work in this.

**Pull request merged**